### PR TITLE
feature(new integration): GA4 cloud mode Integration onboarding

### PR DESCRIPTION
**Pull request by akashrpo**
## Description of the change

> 1. GA4 cloud mode Integration onboarding
> 2. We are using the transformer network layer to send event to GA4
> 3. This PR also includes the bug fix for `constructPayload()` util function where support for handling `false` boolean value is added.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1](rudder-server URL encodes the endpoint before sending an event to the destination and GA4 does not validate if the endpoint is URL encoded due to which we do not see the events being reflected in the dashboard. We have a workaround i.e by utilizing the transformer network layer to send the firebase client event to GA4)

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/ga4/config.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+const { getMappingConfig } = require("../../util");
+
+const ENDPOINT = "https://www.google-analytics.com/mp/collect";
+
+const ConfigCategory = {
```
Comment (utsabc) [inline v0/destinations/ga4/config.js:5]: Can we add reference to Rudderstack ecom spec for each of these events

File: v0/destinations/ga4/data/GA4CommonConfig.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
+[
+  {
+    "destKey": "client_id",
+    "sourceKeys": "messageId",
```
Comment (utsabc) [inline v0/destinations/ga4/data/GA4CommonConfig.json:4]: lets add support for some custom id as well for this`context.client_id || messageId`

File: v0/destinations/ga4/data/GA4PostScoreConfig.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "required": true
+  },
+  {
+    "destKey": "level",
+    "sourceKeys": "properties.level",
+    "required": false
+  },
+  {
+    "destKey": "character",
+    "sourceKeys": "properties.charcharacteracter",
```
Comment (utsabc) [inline v0/destinations/ga4/data/GA4PostScoreConfig.json:14]: There is typo here probably

File: v0/destinations/ga4/transform.js [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+  };
+  response.params = {
+    api_secret: Config.apiSecret,
+    measurement_id: Config.measurementId
+  };
+  response.body.JSON = rawPayload;
+  return response;
+};
+
+const process = async event => {
```
Comment (utsabc) [inline v0/destinations/ga4/transform.js:358]: do we need async here?

File: v0/destinations/ga4/transform.js [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    delete rawPayload.client_id;
+  }
+
+  const payload = {};
+  if (eventNameMapping[event.toLowerCase()]) {
+    // GA4 standard events
+    // get event specific parameters
+    switch (event.toLowerCase()) {
+      /* Browsing Section */
+      case "products_searched":
```
Comment (utsabc) [inline v0/destinations/ga4/transform.js:85]: Based on this ref:https://www.rudderstack.com/docs/rudderstack-api/api-specification/rudderstack-ecommerce-events-specification/ordering/

We shouldn't consider only snake case event naming here

Comment (akashrpo) [reply v0/destinations/ga4/transform.js:85]: We are bringing strings to a common ground by trimming and replacing spaces with `_`. [L:40](https://github.com/rudderlabs/rudder-transformer/pull/1130/files/2b1d183dbf6234b61d1a047e1285e3259e74987c#diff-38fd1bc16aadc482899b425f3baf343c7ccfc49e7a9f0b794bae8f79a4ffaf04R40)

File: v0/destinations/ga4/data/GA4SpendVirtualCurrencyConfig.json [hunk 6]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (ItsSudip) [inline v0/destinations/ga4/data/GA4SpendVirtualCurrencyConfig.json:1]: Why it is empty? Is it added due to remaining consistent, like for other events we are using a config.json we should use one for this event also?

File: v0/destinations/ga4/transform.js [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+const trackResponseBuilder = async (message, { Config }) => {
+  let event = get(message, "event");
+  if (!event) {
+    throw new CustomError("Event name is required", 400);
+  }
+
+  // trim and replace spaces with "_"
+  event = event.trim().replace(/\s+/g, "_");
+
```
Comment (ItsSudip) [inline v0/destinations/ga4/transform.js:41]: Apart from calling lowerCase every time we can make the event lowercase here. WDYT?

File: v0/destinations/ga4/utils.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    items = [];
+    products.forEach((item, index) => {
+      let element = constructPayload(
+        item,
+        mappingConfig[ConfigCategory.ITEMS.name]
+      );
+      if (!element.item_id && !element.item_name) {
+        throw new CustomError("One of product_id or name is required", 400);
+      }
+
```
Comment (ItsSudip) [inline v0/destinations/ga4/utils.js:228]: Don't discard the entire event for a single product.

File: v0/destinations/ga4/transform.js [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+    case EventType.TRACK:
+      response = await trackResponseBuilder(message, destination);
+      break;
+    default:
+      throw new CustomError(`Message type ${messageType} not supported`, 400);
+  }
+  return response;
+};
+
+module.exports = { process };
```
Comment (ItsSudip) [inline v0/destinations/ga4/transform.js:414]: Can't we support router transformation?

**Comment by ItsSudip** [comment]
Suggestion: 
1. Add more unit test cases.
2. Double check all the event names are properly mapped as well as config.json files.

File: v0/destinations/ga4/transform.js (hunk 7 above)
Comment (akashrpo) [reply v0/destinations/ga4/transform.js:41]: Due to custom events being case sensitive

File: v0/destinations/ga4/utils.js (hunk 8 above)
Comment (akashrpo) [reply v0/destinations/ga4/utils.js:228]: It is a required field. [Ref](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events#begin_checkout_item)

**Comment by akashrpo** [comment]
Test cases are yet to be added after testing each component.

File: v0/destinations/ga4/transform.js (hunk 9 above)
Comment (akashrpo) [reply v0/destinations/ga4/transform.js:414]: we are planning to use CDK for GA4 later due to which we decided not to do routerTransform for now.

File: v0/destinations/ga4/utils.js (hunk 8 above)
Comment (ItsSudip) [reply v0/destinations/ga4/utils.js:228]: Yes, that I understood. My concern is we are discarding the entire call for a single product being missed the required field. Can't we implement in that way, filer out the wrong products and send the correct products only?

Comment (akashrpo) [reply v0/destinations/ga4/utils.js:228]: If we do this then the filtration will be done silently without the user being aware of the properties they are trying to send. IMO keeping this gives the user a better idea of what exactly needs to be passed to get proper reporting on the GA4 dashboard. cc @utsabc

File: v0/destinations/ga4/data/GA4SpendVirtualCurrencyConfig.json (hunk 6 above)
Comment (akashrpo) [reply v0/destinations/ga4/data/GA4SpendVirtualCurrencyConfig.json]: Added. It got left out.

File: v0/destinations/ga4/data/GA4GroupConfig.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "destKey": "group_id",
+    "sourceKeys": "properties.group_id",
```
Comment (ItsSudip) [inline v0/destinations/ga4/data/GA4GroupConfig.json:4]: We can get groupId in the root of the message.

File: v0/destinations/ga4/data/GA4ItemsConfig.json [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  },
+  {
+    "destKey": "index",
+    "sourceKeys": "position",
+    "metadata": {
+      "type": "toNumber"
+    },
+    "required": false
+  }
+]
```
Comment (ItsSudip) [inline v0/destinations/ga4/data/GA4ItemsConfig.json:56]: Can't see anything related to this file except it is mentioned once in the config.js.

Comment (akashrpo) [reply v0/destinations/ga4/data/GA4ItemsConfig.json:56]: we are using it to get items properties and is used in utils.js. `getDestinationItemProperties()`

File: v0/destinations/ga4/transform.js (hunk 4 above)
Comment (utsabc) [reply v0/destinations/ga4/transform.js:409]: async doesn't make sense if you are not using promise inside a function. The feature you are mentioning its there in versionedRouter js where we make async processing of events
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

Comment (akashrpo) [reply v0/destinations/ga4/transform.js]: removed

File: v0/destinations/ga4/utils.js (hunk 8 above)
Comment (utsabc) [reply v0/destinations/ga4/utils.js:228]: I agree with @akashrpo on this but !element.item_id will not work when item id is `0` we should use defined and not null

Comment (akashrpo) [reply v0/destinations/ga4/utils.js:232]: Added this check

**Review by ItsSudip — DISMISSED** [review]

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**